Handle unqualified imports in custom Triton kernels

aakhundov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123703

Summary: If in a custom (user-written) Triton kernel an externally imported symbol is used directly, we need to codegen the corresponding import outside the kernel body in the Python wrapper. E.g., if the user code has this:

```
    from triton.language.extra.cuda.libdevice import fast_dividef

    @triton.jit
    def my_kernel(...):
        ...
        x = fast_dividef(...)
        ...
```

The `from triton.language.extra.cuda.libdevice import fast_dividef` line needs to be carried over together with the `my_kernel` function. The PR adds this.

Test Plan:

```
$ python test/inductor/test_triton_kernels.py
...
----------------------------------------------------------------------
Ran 464 tests in 113.512s

OK
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D55953241](https://our.internmc.facebook.com/intern/diff/D55953241)

aakhundov committed:
Handle unqualified imports in custom Triton kernels

Summary: If in a custom (user-written) Triton kernel an externally imported symbol is used directly, we need to codegen the corresponding import outside the kernel body in the Python wrapper. E.g., if the user code has this:

```
    from triton.language.extra.cuda.libdevice import fast_dividef

    @triton.jit
    def my_kernel(...):
        ...
        x = fast_dividef(...)
        ...
```

The `from triton.language.extra.cuda.libdevice import fast_dividef` line needs to be carried over together with the `my_kernel` function. The PR adds this.

Test Plan:

```
$ python test/inductor/test_triton_kernels.py
...
----------------------------------------------------------------------
Ran 464 tests in 113.512s

OK
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

aakhundov commented:
@aakhundov has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55953241).

aakhundov committed:
Update on "Handle unqualified imports in custom Triton kernels"


Summary: If in a custom (user-written) Triton kernel an externally imported symbol is used directly, we need to codegen the corresponding import outside the kernel body in the Python wrapper. E.g., if the user code has this:

```
    from triton.language.extra.cuda.libdevice import fast_dividef

    triton.jit
    def my_kernel(...):
        ...
        x = fast_dividef(...)
        ...
```

The `from triton.language.extra.cuda.libdevice import fast_dividef` line needs to be carried over together with the `my_kernel` function. The PR adds this.

Test Plan:

```
$ python test/inductor/test_triton_kernels.py
...
----------------------------------------------------------------------
Ran 464 tests in 113.512s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

Differential Revision: [D55953241](https://our.internmc.facebook.com/intern/diff/D55953241)

[ghstack-poisoned]

aakhundov commented:
@aakhundov has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55953241).

oulgen commented on test/inductor/test_triton_kernels.py:
```diff
@@ -27,6 +27,11 @@
     import triton
     from triton import language as tl
 
+    from triton.language.extra.cuda.libdevice import (
```
do you need to import this at top level? could you do it inside the particular test?

aakhundov replied:
No, Triton compiler can't pick it up when the import is inside the test, unfortunately.

oulgen commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1150,6 +1151,14 @@ def define_user_defined_triton_kernel(self, kernel, configs, kwargs):
+                node.id
+                for node in ast.walk(ast.parse(cur_kernel.src))
+                if isinstance(node, ast.Name)
+                and hasattr(node, "ctx")
```
add some comments here

how is

import foo
from bar import baz as fizz

is seen by this ast walk

aakhundov replied:
This doesn't actually see any imports, as those are outside the kernel code. What this does is fetching all names "loaded" directly within the code of the kernel. And then we match those to the names in the `__globals__`. Will add a comment.

oulgen commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1165,6 +1174,17 @@ def traverse(cur_kernel):
+                    ):
+                        # a global symbol referenced without module
+                        # qualification: need to codegen an import
+                        compile_wrapper.writeline(
```
Hmm perhaps we should have a "safe" list of things we can emit here? otherwise we will end up executing import time deps of unknown code.

wdyt?

aakhundov replied:
We can limit to the module names starting from `triton.` maybe? As other things are not usable in Triton kernels anyway. Do you think that would be safe enough? One thing is that, if something unqualified is *used* in the kernel code and exists in globals and has a `__module__`, we likely should import it for the kernel to be compileable.

oulgen commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1150,6 +1151,14 @@ def define_user_defined_triton_kernel(self, kernel, configs, kwargs):
         symbols_included = {original_name}
 
         def traverse(cur_kernel):
+            unqualified_loads = {
```
if two kernels are in the same file, you'll parse it twice here

aakhundov replied:
That's true: for every kernel in the file we'll do this exactly once. But we won't emit two imports of the same symbol used in two different kernels (if that's what you mean), as the symbol will be added to the `symbols_included` set on first import emission?

aakhundov committed:
Update on "Handle unqualified imports in custom Triton kernels"


Summary: If in a custom (user-written) Triton kernel an externally imported symbol is used directly, we need to codegen the corresponding import outside the kernel body in the Python wrapper. E.g., if the user code has this:

```
    from triton.language.extra.cuda.libdevice import fast_dividef

    triton.jit
    def my_kernel(...):
        ...
        x = fast_dividef(...)
        ...
```

The `from triton.language.extra.cuda.libdevice import fast_dividef` line needs to be carried over together with the `my_kernel` function. The PR adds this.

Test Plan:

```
$ python test/inductor/test_triton_kernels.py
...
----------------------------------------------------------------------
Ran 464 tests in 113.512s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

Differential Revision: [D55953241](https://our.internmc.facebook.com/intern/diff/D55953241)

[ghstack-poisoned]

aakhundov committed:
Update on "Handle unqualified imports in custom Triton kernels"


Summary: If in a custom (user-written) Triton kernel an externally imported symbol is used directly, we need to codegen the corresponding import outside the kernel body in the Python wrapper. E.g., if the user code has this:

```
    from triton.language.extra.cuda.libdevice import fast_dividef

    triton.jit
    def my_kernel(...):
        ...
        x = fast_dividef(...)
        ...
```

The `from triton.language.extra.cuda.libdevice import fast_dividef` line needs to be carried over together with the `my_kernel` function. The PR adds this.

Test Plan:

```
$ python test/inductor/test_triton_kernels.py
...
----------------------------------------------------------------------
Ran 464 tests in 113.512s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

Differential Revision: [D55953241](https://our.internmc.facebook.com/intern/diff/D55953241)

[ghstack-poisoned]

aakhundov commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1150,6 +1151,18 @@ def define_user_defined_triton_kernel(self, kernel, configs, kwargs):
+            # the respective imports necessary for the kernel compilation
+            unqualified_loads = {
+                node.id
+                for node in ast.walk(ast.parse(cur_kernel.src))
```
@jansel does it sound fine to you to use `ast` here? The purpose is extracting unqualified names from the kernel code as candidates for the globals imported externally. Thanks!

jansel replied:
I think this would be a better way to do the same thing:
```python
>>> {inst.argval for inst in dis.Bytecode(add_kernel_with_imported_symbol)
     if inst.opcode == dis.LOAD_GLOBAL}
{'tl', 'fast_dividef'}
```

aakhundov commented:
@aakhundov has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55953241).

aakhundov committed:
Update on "Handle unqualified imports in custom Triton kernels"


Summary: If in a custom (user-written) Triton kernel an externally imported symbol is used directly, we need to codegen the corresponding import outside the kernel body in the Python wrapper. E.g., if the user code has this:

```
    from triton.language.extra.cuda.libdevice import fast_dividef

    triton.jit
    def my_kernel(...):
        ...
        x = fast_dividef(...)
        ...
```

The `from triton.language.extra.cuda.libdevice import fast_dividef` line needs to be carried over together with the `my_kernel` function. The PR adds this.

Test Plan:

```
$ python test/inductor/test_triton_kernels.py
...
----------------------------------------------------------------------
Ran 464 tests in 113.512s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

Differential Revision: [D55953241](https://our.internmc.facebook.com/intern/diff/D55953241)

[ghstack-poisoned]

aakhundov commented:
@aakhundov has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55953241).

aakhundov committed:
Update on "Handle unqualified imports in custom Triton kernels"


Summary: If in a custom (user-written) Triton kernel an externally imported symbol is used directly, we need to codegen the corresponding import outside the kernel body in the Python wrapper. E.g., if the user code has this:

```
    from triton.language.extra.cuda.libdevice import fast_dividef

    triton.jit
    def my_kernel(...):
        ...
        x = fast_dividef(...)
        ...
```

The `from triton.language.extra.cuda.libdevice import fast_dividef` line needs to be carried over together with the `my_kernel` function. The PR adds this.

Test Plan:

```
$ python test/inductor/test_triton_kernels.py
...
----------------------------------------------------------------------
Ran 464 tests in 113.512s

OK
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

Differential Revision: [D55953241](https://our.internmc.facebook.com/intern/diff/D55953241)

[ghstack-poisoned]

aakhundov commented:
@aakhundov has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55953241).

aakhundov commented:
For posterity: switched to `inst.opname == "LOAD_GLOBAL"` due to the `dis.LOAD_GLOBAL` failing internally due to the missing symbol.

aakhundov commented:
@pytorchbot merge
